src/handlers/keyboard_handler.rs

```
@@ -0,0 +1,460 @@
+        _: u32,
+    ) {
+        if let Some(active_surface) = &mut self.active_keyboard_surface {
+            if active_surface.id() == wl_surface.id() {
+                self.active_keyboard_surface.take();
+            }
+        }
+    }
+
+    fn press_key(
+        &mut self,
+        _: &smithay_client_toolkit::reexports::client::Connection,
+        _: &smithay_client_toolkit::reexports::client::QueueHandle<Self>,
+        _: &smithay_client_toolkit::reexports::client::protocol::wl_keyboard::WlKeyboard,
+        _: u32,
+        event: smithay_client_toolkit::seat::keyboard::KeyEvent,
+    ) {
+        if let Some(active_surface) = &self.active_keyboard_surface {
+            let smithay_windows = self.world().non_send_resource::<SmithayWindows>();
+            let entity = smithay_windows
+                .smithay_to_entity
+                .get(&active_surface.id())
+                .expect("no window created for the keyboard surface!");
+
+            let bevy_event = convert_keyboard_event(event, *entity, ButtonState::Pressed);
+            self.bevy_window_events.send(bevy_event);
+        } else {
+            panic!("There is no window available to send keyboard events!");
+        }
+    }
+
+    fn release_key(
+        &mut self,
+        _: &smithay_client_toolkit::reexports::client::Connection,
+        _: &smithay_client_toolkit::reexports::client::QueueHandle<Self>,
+        _: &smithay_client_toolkit::reexports::client::protocol::wl_keyboard::WlKeyboard,
+        _: u32,
+        event: smithay_client_toolkit::seat::keyboard::KeyEvent,
+    ) {
+        if let Some(active_surface) = &self.active_keyboard_surface {
+            let smithay_windows = self.world().non_send_resource::<SmithayWindows>();
+            let entity = smithay_windows
+                .smithay_to_entity
+                .get(&active_surface.id())
+                .expect("no window created for the keyboard surface!");
```

Instead of expect or panic, should we soft ignore this scenario and throw a warning?

---

Yep that should be the way to go.